File: AoC/2023_problems/day1.py
test = """1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet"""


test2 = """two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen"""


def solve(input, part):
    total = 0
    for line in input.split():
        if part == 2:
            line = fixline(line)
        for c in line:
            if c.isnumeric():
                total += int(c) * 10
                break
        for c in line[::-1]:
            if c.isnumeric():
                total += int(c)
                break
    print(total)


def fixline(line):
    nums = {
        "one": 1,
        "two": 2,
        "three": 3,
        "four": 4,
        "five": 5,
        "six": 6,
        "seven": 7,
        "eight": 8,
        "nine": 9
    }
    # check from front
    sub = ""
    for c in line:
        if c.isnumeric():
            break
        sub += c
        for n in nums:
            if n in sub:
                line = line.replace(n, str(nums[n]))
                break
    # check from back
    sub = ""
    for c in line[::-1]:
        if c.isnumeric():
            break
        sub = c + sub
        for n in nums:
            if n in sub:
                line = line.replace(n, str(nums[n]))
                break
    return line


with open("AoC/2023_problems/day1.txt", "r") as f:
    inpt = f.read()
    solve(inpt, 1)
    solve(inpt, 2)

File: AoC/2023_problems/day1.js
const fs = require("fs");

const data = fs.readFileSync("AoC/2023_problems/day1.txt", "utf8");

// Gets the two digit number from a line by finding first and last digit
function getNumber(line) {
    let nums = [];
    for (const char of line) {
        if (!isNaN(char)) {
            nums.push(char);
        }
    }
    if (nums.length > 1) {
        return parseInt(nums[0] + nums[nums.length - 1]);
    }
    return parseInt(nums[0] + nums[0]);
}

// Replaces name of a number with a string containg that number.
// We avoid chaning first and last letter since it can be a part of another number
// oneight
function replaceWords(line) {
    const numMap = {
        "one": "o1e",
        "two": "t2o",
        "three": "thr3e",
        "four": "fo4r",
        "five": "fi5e",
        "six": "s6x",
        "seven": "se7en",
        "eight": "ei8ht",
        "nine": "ni9e"
    };
    for (const [key, value] of Object.entries(numMap)) {
        line = line.replaceAll(key, value);
    }
    return line;
}

function solution(data) {
    const lines = data.split("\n");
    let part1 = 0;
    let part2 = 0
    for (const line of lines) {
        part1 += getNumber(line);
        part2 += getNumber(replaceWords(line));
    }
    return [part1, part2];
}

console.log(solution(data));
